File: frontend/src/types/todo-attachment.ts
// TypeScript types for todo file attachments

export interface TodoAttachment {
  id: number;
  todo_id: number;
  original_filename: string;
  file_size: number;
  mime_type: string;
  file_extension?: string;
  description?: string;
  uploaded_by?: number;
  uploaded_at: string;
  updated_at: string;
  
  // Computed properties
  file_size_mb: number;
  file_category: 'image' | 'document' | 'video' | 'audio' | 'other';
  file_icon: string;
  download_url: string;
  preview_url?: string;
  is_image: boolean;
  is_document: boolean;
  is_video: boolean;
  is_audio: boolean;
}

export interface TodoAttachmentList {
  attachments: TodoAttachment[];
  total_count: number;
  total_size_mb: number;
}

export interface FileUploadResponse {
  message: string;
  attachment: TodoAttachment;
}

export interface FileUploadRequest {
  description?: string;
}

export interface AttachmentStats {
  total_attachments: number;
  total_size_mb: number;
  file_type_counts: Record<string, number>;
  largest_file?: TodoAttachment;
  recent_attachments: TodoAttachment[];
}

export interface BulkDeleteRequest {
  attachment_ids: number[];
}

export interface BulkDeleteResponse {
  message: string;
  deleted_count: number;
  failed_deletions: Array<{
    attachment_id: number;
    filename: string;
    error: string;
  }>;
}

// File upload progress tracking
export interface FileUploadProgress {
  file: File;
  progress: number;
  status: 'pending' | 'uploading' | 'success' | 'error';
  error?: string;
  attachment?: TodoAttachment;
}

// File validation result
export interface FileValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

// Upload configuration
export interface UploadConfig {
  maxFileSize: number; // in bytes (25MB = 26214400)
  allowedTypes?: string[]; // MIME types, if empty = all types allowed
  maxFiles?: number; // max files per todo, if empty = unlimited
}

// File category icons mapping
export const FILE_CATEGORY_ICONS = {
  image: 'PhotoIcon',
  document: 'DocumentTextIcon', 
  video: 'VideoCameraIcon',
  audio: 'SpeakerWaveIcon',
  other: 'DocumentIcon'
} as const;

// File size formatting utility
export const formatFileSize = (bytes: number): string => {
  if (bytes === 0) return '0 Bytes';
  
  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(2))} ${sizes[i]}`;
};

// File type checking utilities
export const isImageFile = (mimeType: string): boolean => {
  return mimeType.startsWith('image/');
};

export const isDocumentFile = (mimeType: string): boolean => {
  const documentTypes = [
    'application/pdf',
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.ms-excel',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'application/vnd.ms-powerpoint',
    'application/vnd.openxmlformats-officedocument.presentationml.presentation',
    'text/plain',
    'text/csv'
  ];
  return documentTypes.includes(mimeType);
};

export const isVideoFile = (mimeType: string): boolean => {
  return mimeType.startsWith('video/');
};

export const isAudioFile = (mimeType: string): boolean => {
  return mimeType.startsWith('audio/');
};

export const getFileCategory = (mimeType: string): TodoAttachment['file_category'] => {
  if (isImageFile(mimeType)) return 'image';
  if (isDocumentFile(mimeType)) return 'document';
  if (isVideoFile(mimeType)) return 'video';
  if (isAudioFile(mimeType)) return 'audio';
  return 'other';
};

// File validation utilities
export const validateFile = (file: File, config: UploadConfig): FileValidationResult => {
  const result: FileValidationResult = {
    valid: true,
    errors: [],
    warnings: []
  };

  // Check file size
  if (file.size > config.maxFileSize) {
    result.valid = false;
    result.errors.push(
      `File size ${formatFileSize(file.size)} exceeds maximum allowed size of ${formatFileSize(config.maxFileSize)}`
    );
  }

  if (file.size === 0) {
    result.valid = false;
    result.errors.push('File is empty');
  }

  // Check file type if restrictions exist
  if (config.allowedTypes && config.allowedTypes.length > 0) {
    if (!config.allowedTypes.includes(file.type)) {
      result.valid = false;
      result.errors.push(`File type '${file.type}' is not allowed`);
    }
  }

  // Add warnings for large files
  if (file.size > 10 * 1024 * 1024) { // 10MB
    result.warnings.push('Large file may take longer to upload');
  }

  return result;
};

// Default upload configuration
export const DEFAULT_UPLOAD_CONFIG: UploadConfig = {
  maxFileSize: 25 * 1024 * 1024, // 25MB
  allowedTypes: [], // All types allowed
  maxFiles: undefined // No limit
};

// Component Props Interfaces
export interface AttachmentListProps {
  todoId: number;
  attachments: TodoAttachment[];
  onAttachmentDeleted: (attachmentId: number) => void;
  onAttachmentUpdated: (attachment: TodoAttachment) => void;
  className?: string;
  showActions?: boolean;
}

export interface FileUploadProps {
  todoId: number;
  onUploadSuccess: (attachment: TodoAttachment) => void;
  onUploadError: (error: string) => void;
  className?: string;
  disabled?: boolean;
}

File: backend/app/models/todo_attachment.py
from __future__ import annotations

import os
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import BigInteger, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base

if TYPE_CHECKING:
    from app.models.todo import Todo
    from app.models.user import User


class TodoAttachment(Base):
    """Model for todo file attachments with no file count limit and 25MB size limit."""

    __tablename__ = "todo_attachments"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    todo_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("todos.id", ondelete="CASCADE"), nullable=False, index=True
    )
    uploaded_by: Mapped[int] = mapped_column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

    # File information
    original_filename: Mapped[str] = mapped_column(String(255), nullable=False)
    stored_filename: Mapped[str] = mapped_column(String(255), nullable=False, unique=True)
    file_path: Mapped[str] = mapped_column(String(500), nullable=False)
    file_size: Mapped[int] = mapped_column(BigInteger, nullable=False)  # Size in bytes
    mime_type: Mapped[str] = mapped_column(String(100), nullable=False)
    file_extension: Mapped[str | None] = mapped_column(String(10), nullable=True)

    # Optional metadata
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Timestamps
    uploaded_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow, nullable=False, index=True
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
    )

    # Relationships
    todo: Mapped[Todo] = relationship("Todo", backref="attachments")
    uploader: Mapped[User | None] = relationship(
        "User", foreign_keys=[uploaded_by], backref="uploaded_attachments"
    )

    @property
    def file_size_mb(self) -> float:
        """Return file size in megabytes."""
        return self.file_size / (1024 * 1024)

    @property
    def is_image(self) -> bool:
        """Check if the file is an image."""
        image_types = {'image/jpeg', 'image/png', 'image/gif', 'image/webp', 'image/svg+xml'}
        return self.mime_type in image_types

    @property
    def is_document(self) -> bool:
        """Check if the file is a document."""
        doc_types = {
            'application/pdf',
            'application/msword',
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            'application/vnd.ms-excel',
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
            'application/vnd.ms-powerpoint',
            'application/vnd.openxmlformats-officedocument.presentationml.presentation',
            'text/plain',
            'text/csv'
        }
        return self.mime_type in doc_types

    @property
    def is_video(self) -> bool:
        """Check if the file is a video."""
        return self.mime_type.startswith('video/')

    @property
    def is_audio(self) -> bool:
        """Check if the file is audio."""
        return self.mime_type.startswith('audio/')

    @property
    def file_category(self) -> str:
        """Get the file category for UI display."""
        if self.is_image:
            return "image"
        elif self.is_document:
            return "document"
        elif self.is_video:
            return "video"
        elif self.is_audio:
            return "audio"
        else:
            return "other"

    def get_file_icon(self) -> str:
        """Get appropriate icon name for the file type."""
        category = self.file_category
        icon_map = {
            "image": "PhotoIcon",
            "document": "DocumentTextIcon",
            "video": "VideoCameraIcon",
            "audio": "SpeakerWaveIcon",
            "other": "DocumentIcon"
        }
        return icon_map.get(category, "DocumentIcon")

    def get_download_url(self) -> str:
        """Generate download URL for the attachment."""
        return f"/api/todos/{self.todo_id}/attachments/{self.id}/download"

    def get_preview_url(self) -> str | None:
        """Generate preview URL for supported file types."""
        if self.is_image or self.mime_type == 'application/pdf':
            return f"/api/todos/{self.todo_id}/attachments/{self.id}/preview"
        return None

    def is_file_exists(self) -> bool:
        """Check if the physical file exists on disk."""
        return os.path.exists(self.file_path)

    def __repr__(self) -> str:
        return f"<TodoAttachment(id={self.id}, filename='{self.original_filename}', size={self.file_size_mb:.2f}MB)>"
